src/sdl/YamlSDLParser.ts

```diff
@@ -0,0 +1,192 @@
+});
+
+export const v2ServiceDeployment = z.object({
+  profile: z.string(),
+  count: z.number()
+});
+
+export const v2Deployment = z.record(v2ServiceDeployment);
+
+export const v2Endpoint = z.object({
+  kind: z.string()
+});
+
+export const v2Profiles = z.object({
+  compute: z.record(
+    z.object({
+      resources: z.object({
+        cpu: z.object({
+          units: z.union([z.number(), z.string()]),
+          attributes: z.record(z.any()).optional()
+        }),
+        memory: z.object({
+          size: z.string(),
+          attributes: z.record(z.any()).optional()
+        }),
+        storage: z.union([
+          z.array(
+            z.object({
+              name: z.string(),
+              size: z.string(),
+              attributes: z.record(z.any())
+            })
+          ),
+          z.object({
+            name: z.string(),
+            size: z.string(),
+            attributes: z.record(z.any())
+          })
+        ])
+      })
+    })
+  ),
+  placement: z.record(
+    z.object({
+      attributes: z.record(z.string()),
+      signedBy: z.object({
+        allOf: z.array(z.string()).optional(),
+        anyOf: z.array(z.string()).optional()
+      }),
+      pricing: z.record(
+        z.object({
+          denom: z.string(),
+          value: z.number(),
+          amount: z.number()
+        })
+      )
+    })
+  )
+});
+
+const v2Storage = z.object({
+  name: z.string(),
+  size: z.string(),
+  attributes: z.record(z.any())
+});
+
+export const v2ProfileCompute = z.object({
+  resources: z.object({
+    cpu: z.object({
+      units: z.union([z.number(), z.string()]),
+      attributes: z.record(z.any()).optional()
+    }),
+    memory: z.object({
+      size: z.string(),
+      attributes: z.record(z.any()).optional()
+    }),
+    storage: z.union([z.array(v2Storage), v2Storage])
+  })
+});
+
+export const v3ProfileCompute = z.object({
+  resources: z.object({
+    cpu: v2ProfileCompute.shape.resources.shape.cpu,
+    memory: v2ProfileCompute.shape.resources.shape.memory,
+    storage: v2ProfileCompute.shape.resources.shape.storage,
+    gpu: z.object({
+      units: z.union([z.number(), z.string()]),
+      attributes: z.object({
+        vendor: z.record(
+          z.array(
+            z.object({
+              model: z.string(),
+              ram: z.string().optional(),
+              interface: z.string().optional()
+            })
+          )
+        )
+      })
+    }),
+    id: z.number()
+  })
+});
+
+export const v3Profiles = z.object({
+  compute: z.record(v3ProfileCompute),
+  placement: v2Profiles.shape.placement
+});
+
+const Beta2Sdl = z.object({
+  services: z.record(v2Service),
+  deployment: z.record(v2Deployment),
+  endpoints: z.record(v2Endpoint),
+  profiles: v2Profiles
+});
+
+const Beta3Sdl = z.object({
+  services: z.record(v2Service),
+  deployment: z.record(v2Deployment),
+  endpoints: z.record(v2Endpoint),
+  profiles: v3Profiles
+});
+
+export class YamlSDLParser {
```

Comment:
yaml parsing separation simplifies SDL class and allows it to strictly allow inly a parsed SDL on the constructor.